### PR TITLE
CARGO: support items generated by build scripts

### LINKED ISSUE #4631: Support cfg features generated by build scripts
## Environment

* **IntelliJ Rust plugin version:** 0.2.110.5742-193-dev
* **Rust toolchain version:** 1.39 stable
* **IDE name and version:** IntelliJ IDEA 2019.3 beta 2
* **Operating system:** Windows 10 x64

## Problem description
Ctrl-clicking function in specific circumstance takes me to wrong function. Experimental macro expansion engine is enabled.

## Steps to reproduce

Cargo.toml:

```toml
[package]
name = "bug-repro"
version = "0.1.0"
authors = ["Alyosha Vasilieva <1284317+AlyoshaVasilieva@users.noreply.github.com>"]
edition = "2018"

[dependencies]
indexmap = "=1.3.0"
digest = "=0.8.1"
```

main.rs:

```rust
use indexmap::map::IndexMap;

fn main() {
    let mut map = IndexMap::new();
    map.insert("key", "value");
}
```

`IndexMap::new` is highlighted as error with "Unresolved reference: \`new`".

Ctrl-clicking IndexMap, despite being highlighted red, correctly takes me to its declaration.

Ctrl-clicking `new` incorrectly takes me to the `new` function inside `impl<D: Input + FixedOutput + Reset + Clone + Default> Digest for D`, part of the `digest` crate.

Compilation works and gives no warnings or errors.

### PR DESCRIPTION
**Pull request by Undin**
While compilation, crate's [build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) can generate some new code, `cfg` attributes, environment variables and so on.
These changes are the first attempt to support them. What these changes do:
* collect info about executed build scripts while refresh. Under the hood, it uses `cargo check` so it can take much more time to get info about project
* use generated `cfg` options while `cfg` attribute evaluation. Should fix #4631
* use generated environment variables for path evaluation in include macro calls
* extend include macro support. Now, the plugin can process calls without filename in string literals like `include!(env!("BINDINGS"))`. Should make possible to force intelliSence work for `web-sys` (see #3066)
* Since recent nightly builds of cargo, it makes possible to turn on full support of `include` macro without using `cargo build --build-plan` and avoid issue about full project recompilation

Note, it is under experimental feature and disabled by default. To try it just enabled `org.rust.cargo.evaluate.build.scripts` in `Help | Find Action | Experimental Features` dialog

### CODE REVIEW

File: src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt [hunk 1]
```diff
@@ -316,7 +319,7 @@ object CargoMetadata {
             source,
             origin = if (isWorkspaceMember) PackageOrigin.WORKSPACE else PackageOrigin.TRANSITIVE_DEPENDENCY,
             edition = edition.cleanEdition(),
-            features = features
+            features = features + generatedFeatures
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt]: >I don't sure, but probably they are also in different namespaces, so there probably can be a cargo feature and a generated cfg with the same name.

It's rather the other way around: `features` are desugared to cfgs. That is, `feature = "foo"` looks like `--cfg=feature=foo` on the command line. See the relevant desugaring in rust-analyzer: https://github.com/rust-analyzer/rust-analyzer/blob/cdcb3d3833d3d5b37b2cd4dac91a6e9366f20aea/crates/ra_cfg/src/lib.rs#L52-L55

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt]: @matklad Thanks for the suggestion!
I've rewritten code related to generated cfg options. Now it uses `CfgOptions` in a similar way as we support target cfg options like `target_os` and supports key-value generated options.
Probably, we want to merge package feature support with cfg options but I think it should be done in separate PR

File: src/main/kotlin/org/rust/lang/utils/evaluation/CfgEvaluator.kt [hunk 2]
```diff
@@ -84,15 +89,18 @@ class CfgEvaluator(val options: CfgOptions, val features: Map<String, FeatureSta
...
-        name in features -> evaluateFeature(name)
         else -> Unknown
     }
 
-    private fun evaluateNameValue(name: String, value: String): ThreeValuedLogic = when (name) {
-        in SUPPORTED_NAME_VALUE_OPTIONS -> ThreeValuedLogic.fromBoolean(options.isNameValueEnabled(name, value))
+    private fun evaluateNameValue(name: String, value: String): ThreeValuedLogic = when {
+        // TODO: convert whitelist to blacklist and merge options with packageOption
+        name in SUPPORTED_NAME_VALUE_OPTIONS -> ThreeValuedLogic.fromBoolean(options.isNameValueEnabled(name, value))
+        packageOptions.isNameValueEnabled(name, value) -> True
```
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/utils/evaluation/CfgEvaluator.kt]: > Is such a case possible?

Interesting things. And the answer is yes, it's possible (I will refrain from commenting on why it is possible at all).

> If so, do we intentionally prefer options to packageOptions

No, it's not intentional. And looks like we shouldn't prefer `options` or `packageOptions` (I suppose we will merge them in future), we just need to return `true` if it's enabled at list in one place

Nonetheless, the current solution works incorrectly in such cases so I'll fix it. Thanks!

**Comment by Undin** [comment]
bors r=vlad20012

**Comment by mchernyavsky** [comment]
bors retry

**Comment by Undin** [comment]
bors r=vlad20012

**Pull request merged**